text/0000-decouple-dependencies.md

```diff
@@ -0,0 +1,403 @@
+don’t consume buildpacks directly, they consume builders which include
+buildpacks.
+
+This all has to be done as aggressively as possible so that we are shipping
+dependencies, in particular those with security fixes, quickly. This is because
+with metadata in `buildpack.toml`, even if an upstream project releases a bug
+or security fix, buildpack users cannot get that fix until we update and
+release component and composite buildpacks as well as the builder.
+
+There is also toil associated with the tools and pipelines used for this
+process. The tools have bugs and need to be updated. At present, the tools we
+use to manage all of these updates do not scale well either. In particular
+Github Actions, we have had a number of issues hitting rate limits and usage
+caps. This gets worse when there are a lot of dependencies to watch, for
+example, if your buildpack has multiple version lines or different sets of
+packages for a dependency.
+
+The whole process puts an additional maintenance burden on the project
+maintainers and project resources. This is not the type of work that a casual
+contributor to Paketo will do and as we add more dependencies the burden only
+increases on the maintainer teams.
+
+The motivation of this proposal is to…
+
+- Reduce the burden and toil for Paketo buildpack maintainer teams
+- Continue publishing dependency updates in a timely and secure manner
+- Decouple installing dependencies from configuring them
+- Separate metadata and the actual dependencies so they can be provided to
+  buildpacks in a number of different and flexible ways
+- Establish a reasonable release schedule for buildpacks that’s based around
+  development, not dependencies and thus enabling buildpacks to support version
+  lines
+- Make it easier to package buildpacks for offline environments.
+
+## Detailed Explanation
+At a high level:
+
+- We will define a metadata format that includes the metadata fields currently
+  present in `buildpack.toml`.
+- We will add a dependency version validation section to `buildpack.toml`
+  metadata, this can be used to state that a buildpack version only supports
+  certain ranges of a given tool, such as Java `11.*` or Node.js `16.*`.
+- Metadata can be provided by anyone. A user can add custom metadata, or source
+  from a third party project. Paketo will provide an official set of metadata
+  against which we will test the Paketo buildpacks. It will be distributed via
+  images in an image registry (Docker Hub).
+- Buildpacks do not care how dependency metadata is distributed, that is a
+  separate concern, instead, they just read metadata from a specified location.
+  It is on the platform to mount the metadata that it fetched at that location.
+- The actual dependencies are accessed via the metadata and that can happen
+  over any protocol (HTTPS/SFTP/FILE) and be distributed in any format
+  (archive/image).
+- Dependency metadata will be removed from `buildpack.toml`
```

Comment:
I think this RFC should not propose removing dependency metadata from `buildpack.toml`. I think it should describe the ordering of looking for metadata in the new place, followed by the existing `buildpack.toml`. I get that this is an implementation concern of the buildpack, but I think it's worth calling out in the RFC.

I say this because I think it's valuable to provide the ability to fallback to the current mechanism for some period while we incrementally build out the new system and uncover any issues, and I think we should introduce the idea of removing `buildpack.toml` metadata in a separate RFC once all buildpacks have an established model for providing dependencies via the new mechanism.

Reply:
I see that we call out the idea of not removing the current way of doing things in multiple places in the RFC, which is good, so i think probably just need to update this summary bullet point to match the rest of the RFC.

Reply:
+1 removal of the metadata for dependencies in the buildpacks is not part of this RFC. Probably a separate RFC.

Reply:
```suggestion
- Dependency metadata will **not** be removed from `buildpack.toml`, this will be a matter of a separate RFC once all buildpacks have adopted this RFC.
```

FWIW I am not sure if fallback is the best approach. We should at least consider if it's best to error out if for example the buildpack and the user provide metadata - this might well be a conscious choice of the buildpack team to not (yet) support this RFC.

Reply:
I am more than happy pushing that off into another RFC.

Reply:
Let's do that @ForestEckhardt.

As far as fallback or not fallback, I had thought we could make that the user's choice. Like @robdimsdale mentioned, as buildpack authors we have to have a way to slowly move over. So my thought was having an agreed upon env variable, like `BP_EXTERNAL_DEPENDENCIES_DISABLED`, which initially defaults to opt-in, then changes to opt-out, then goes away with the internal metadata. That way we can phase things in nicely.

That said, I think that's part of a follow-up RFC. We can decide upon that as part of deciding on how we deprecate stuff.

Sorry if this RFC seems like it's punting a lot of stuff into other future RFCs. This is kind of like a meta RFC, it's setting up the framework to make this change. We'll need a few more RFCs to fully spell everything out.

Reply:
No problem at all with the concept of meta/top-level RFC. It's what we did with the dependencies rewrite and it worked well.

Reply:
I am a little confused on the direction we want here. Do we want to add a trigger mechanism as part of **THIS** RFC or do we want to hold out for an RFC that is more specific on the implementation? It is sounding like we want to add a mechanism to opt-in and then in a subsequent RFC it will be changed to be an opt-out trigger. Is that correct?

Reply:
Ok while double checking the RFC I think that this use case is covered by `BP_EXTERNAL_METADATA_ENABLED`. Please check out the definition of this environment variable in the implementation section and see if that is what you are looking for.

Reply:
Ah, right. I forgot about that. I think that covers my concern.